### ISSUE TITLE
bedtools compatibility issues

### ISSUE THREAD

**Issue by photocyte** [mpdunne/omgene#2]
Hi there,

I'm trying to run omgene, but it seems that the embedded bedtools commands are out of date (My install has bedtools v2.27.1 - the most recent version).  For example, when I run omgene on my .tsv file, it works for a time then produces this error:

```
python2 /lab/solexa_weng/testtube/omgene/omgene.py -i test.tsv
loading and checking input data locations...
Grabbing cds and aa sequences for inputted transcripts...
Traceback (most recent call last):
  File "/lab/solexa_weng/testtube/omgene/omgene.py", line 3421, in <module>
    go(path_inf, path_ref, path_resultsDir, path_wDir, minintron, minexon, int_numCores, int_slopAmount)
  File "/lab/solexa_weng/testtube/omgene/omgene.py", line 3318, in go
    dict_generegions = prepareGeneregions(dict_seqInfo, dict_genomeInfo, path_wDir, int_numCores,int_slopAmount)#qe
  File "/lab/solexa_weng/testtube/omgene/omgene.py", line 1111, in prepareGeneregions
    path_generegion	= writeGeneRegionFile(line, generegion, path_mDir_l + "/" + str(generegion) + ".gtf")
  File "/lab/solexa_weng/testtube/omgene/omgene.py", line 1082, in writeGeneRegionFile
    ".", line[3], ".", "transcript_id \"" + generegion + "\"; gene_id \"" + generegion + "\"", generegion]
IndexError: list index out of range
```

The "line" variable referenced seems to be the output of a bedtools stream from a previous command, and indeed if I print(line), it has only 3 columns.  This patch (https://github.com/photocyte/omgene/commit/9f48d5fdcc086f42a96d0ceff04efa1ac3436e73) that uses the -c and -o parameters of bedtools seems to fix it, but I'm not sure I implemented it properly, as the script crashes later on with this problem:
```
python2 /lab/solexa_weng/testtube/omgene/omgene.py -i test.tsv
Checking installed programs...
loading and checking input data locations...
Grabbing cds and aa sequences for inputted transcripts...
Traceback (most recent call last):
  File "/lab/solexa_weng/testtube/omgene/omgene.py", line 3421, in <module>
    go(path_inf, path_ref, path_resultsDir, path_wDir, minintron, minexon, int_numCores, int_slopAmount)
  File "/lab/solexa_weng/testtube/omgene/omgene.py", line 3318, in go
    dict_generegions = prepareGeneregions(dict_seqInfo, dict_genomeInfo, path_wDir, int_numCores,int_slopAmount)#qe
  File "/lab/solexa_weng/testtube/omgene/omgene.py", line 1151, in prepareGeneregions
    dg["cdsBase_data"] = str(readSeqs(dg["cdsBase"])[0].seq)
IndexError: list index out of range
```

Any ideas?

All the best,
-Tim

**Comment by photocyte** [comment]
For what its worth, using an older version of bedtools (v2.19.1) does seem to get around the 1st issue, but there is a new issue with the files that are output by the embedded bash script (the files do get seemingly produced, but they are empty).  Then, the 2nd issue from above happens again...

```
python2 /lab/solexa_weng/testtube/omgene/omgene.py -i test.tsv
Checking installed programs...
loading and checking input data locations...
Grabbing cds and aa sequences for inputted transcripts...
Error: The requested file (/tmp/tmp.dLPsxC6N0Q/gtfBed.pos) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.bSfSgnsKmc/gtfBed.pos) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.AJX8WI0vQT/gtfBed.neg) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.KWhhmsyILb/gtfBed.neg) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.K74EOqLvUZ/gtfBed.pos) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.16TbgGMbf8/gtfBed.pos) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.ur0CTvDECi/gtfBed.pos) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.8zGXKHM7f1/gtfBed.pos) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.49nhK43t6X/gtfBed.neg) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.LbRrd5aCSo/gtfBed.neg) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.tbKC7T5JBw/gtfBed.neg) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.3CxEDnHPdd/gtfBed.neg) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.ise4FuKuji/gtfBed.neg) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.hvqfc7fken/gtfBed.neg) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.gXg7Jf8lCR/gtfBed.pos) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.IrdLFSckiL/gtfBed.pos) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.EYu6hHz0iL/gtfBed.neg) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.EYu6hHz0iL/gtfBed.pos) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.bp86JQC1VL/gtfBed.neg) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.bp86JQC1VL/gtfBed.pos) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.suC3W3HdCL/gtfBed.neg) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.suC3W3HdCL/gtfBed.pos) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.ZFnswtVahO/gtfBed.neg) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.ZFnswtVahO/gtfBed.pos) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.I4mtAi6blb/gtfBed.neg) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.I4mtAi6blb/gtfBed.pos) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.RlcjIfc9Js/gtfBed.neg) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.RlcjIfc9Js/gtfBed.pos) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.T31pbTelra/gtfBed.neg) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.T31pbTelra/gtfBed.pos) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.wh2vBhxFpQ/gtfBed.neg) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.wh2vBhxFpQ/gtfBed.pos) could not be opened. Error message: (Success). Exiting!
Traceback (most recent call last):
  File "/lab/solexa_weng/testtube/omgene/omgene.py", line 3421, in <module>
    go(path_inf, path_ref, path_resultsDir, path_wDir, minintron, minexon, int_numCores, int_slopAmount)
  File "/lab/solexa_weng/testtube/omgene/omgene.py", line 3318, in go
    dict_generegions = prepareGeneregions(dict_seqInfo, dict_genomeInfo, path_wDir, int_numCores,int_slopAmount)#qe
  File "/lab/solexa_weng/testtube/omgene/omgene.py", line 1151, in prepareGeneregions
    dg["cdsBase_data"] = str(readSeqs(dg["cdsBase"])[0].seq)
IndexError: list index out of range
```

**Comment by mpdunne** [comment]
[https://tr.cloudmagic.com/h/v6/emailtag/tag/2.0/1529521098/362f86978213108f86a5a8b78262f84e/3/c27e319654c1956ad6e4517247af6c32/714a6f60a31a99707be5d8d004a7f111/9efab2399c7c560b34de477b9aa0a465/newton.gif]

Hi Tim!

Sorry you've been having these issues, and thanks for flagging them up. I'll look into them asap!

All the best,

Michael


On Wed, Jun 20, 2018 at 7:40pm, Timothy R. Fallon <notifications@github.com<mailto:notifications@github.com>> wrote:


For what its worth, using an older version of bedtools (v2.19.1) does seem to get around the 1st issue, but there is a new issue with the files that are output by the embedded bash script (the files do get seemingly produced, but they are empty). Then, the 2nd issue from above happens again...

python2 /lab/solexa_weng/testtube/omgene/omgene.py -i test.tsv
Checking installed programs...
loading and checking input data locations...
Grabbing cds and aa sequences for inputted transcripts...
Error: The requested file (/tmp/tmp.dLPsxC6N0Q/gtfBed.pos) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.bSfSgnsKmc/gtfBed.pos) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.AJX8WI0vQT/gtfBed.neg) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.KWhhmsyILb/gtfBed.neg) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.K74EOqLvUZ/gtfBed.pos) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.16TbgGMbf8/gtfBed.pos) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.ur0CTvDECi/gtfBed.pos) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.8zGXKHM7f1/gtfBed.pos) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.49nhK43t6X/gtfBed.neg) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.LbRrd5aCSo/gtfBed.neg) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.tbKC7T5JBw/gtfBed.neg) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.3CxEDnHPdd/gtfBed.neg) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.ise4FuKuji/gtfBed.neg) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.hvqfc7fken/gtfBed.neg) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.gXg7Jf8lCR/gtfBed.pos) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.IrdLFSckiL/gtfBed.pos) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.EYu6hHz0iL/gtfBed.neg) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.EYu6hHz0iL/gtfBed.pos) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.bp86JQC1VL/gtfBed.neg) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.bp86JQC1VL/gtfBed.pos) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.suC3W3HdCL/gtfBed.neg) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.suC3W3HdCL/gtfBed.pos) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.ZFnswtVahO/gtfBed.neg) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.ZFnswtVahO/gtfBed.pos) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.I4mtAi6blb/gtfBed.neg) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.I4mtAi6blb/gtfBed.pos) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.RlcjIfc9Js/gtfBed.neg) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.RlcjIfc9Js/gtfBed.pos) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.T31pbTelra/gtfBed.neg) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.T31pbTelra/gtfBed.pos) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.wh2vBhxFpQ/gtfBed.neg) could not be opened. Error message: (Success). Exiting!
Error: The requested file (/tmp/tmp.wh2vBhxFpQ/gtfBed.pos) could not be opened. Error message: (Success). Exiting!
Traceback (most recent call last):
  File "/lab/solexa_weng/testtube/omgene/omgene.py", line 3421, in <module>
    go(path_inf, path_ref, path_resultsDir, path_wDir, minintron, minexon, int_numCores, int_slopAmount)
  File "/lab/solexa_weng/testtube/omgene/omgene.py", line 3318, in go
    dict_generegions = prepareGeneregions(dict_seqInfo, dict_genomeInfo, path_wDir, int_numCores,int_slopAmount)#qe
  File "/lab/solexa_weng/testtube/omgene/omgene.py", line 1151, in prepareGeneregions
    dg["cdsBase_data"] = str(readSeqs(dg["cdsBase"])[0].seq)
IndexError: list index out of range

**Comment by mpdunne** [comment]
Hi Tim!

I've looked into the merging problem. It turns out the output and parameters for bedtools merge from several different versions are all different (I tried 2.19, 2.26, and 2.27). The issue relates specifically to strand-specific merging. As a safer solution, I've implemented a fix which separates the strands before merging, and then recombines them.

The second issue is separate. Without seeing your inputted GTF I can't be sure, but I suspect OMGene is expecting upper-case "CDS" tags in the third column, and the inputted GTF file has "cds". I've updated the code to ignore case for this tag. If this is not the issue, if you could send over your GTF file I'll have another look.

Let me know if there's anything else!

All the best,

Michael

**Comment by photocyte** [comment]
Hi Michael, thanks a bunch! I've pulled the most recent version & with some changes (https://github.com/mpdunne/omgene/pull/3) it seems to be working well.  Yes, my .GTF files had an uppercase CDS attribute.  Good catch.

Of course, after successful execution none of my gene models were changed, but that is more likely due to the fact that the gene models were already okay.  Will try testing with a more problematic gene family.

All the best,
-Tim
